kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DomainMetadataSuite.scala

```diff
@@ -0,0 +1,475 @@
+    // Create an empty table
+    createTxn(engine, tablePath, isNewTable = true, testSchema, Seq.empty)
+      .commit(engine, emptyIterable())
+
+    // Set writer version and writer feature to support domain metadata
+    setDomainMetadataSupport(engine, tablePath)
+  }
+
+  private def validateDomainMetadataConflictResolution(
+      engine: Engine,
+      tablePath: String,
+      currentTxn1DomainMetadatas: Seq[DomainMetadata],
+      winningTxn2DomainMetadatas: Seq[DomainMetadata],
+      winningTxn3DomainMetadatas: Seq[DomainMetadata],
+      expectedConflict: Boolean): Unit = {
+    // Create table with domain metadata support
+    createTableWithDomainMetadataSupported(engine, tablePath)
+    val table = Table.forPath(engine, tablePath)
+
+    /**
+     * Txn1: i.e. the current transaction that comes later than winning transactions.
+     * Txn2: i.e. the winning transaction that was committed first.
+     * Txn3: i.e. the winning transaction that was committed secondly.
+     *
+     * Note tx is the timestamp.
+     *
+     * t1 ------------------------ Txn1 starts.
+     * t2 ------- Txn2 starts.
+     * t3 ------- Txn2 commits.
+     * t4 ------- Txn3 starts.
+     * t5 ------- Txn3 commits.
+     * t6 ------------------------ Txn1 commits (SUCCESS or FAIL).
+     */
+    val txn1 = createTxnWithDomainMetadatas(engine, tablePath, currentTxn1DomainMetadatas)
+
+    val txn2 = createTxnWithDomainMetadatas(engine, tablePath, winningTxn2DomainMetadatas)
+    txn2.commit(engine, emptyIterable())
+
+    val txn3 = createTxnWithDomainMetadatas(engine, tablePath, winningTxn3DomainMetadatas)
+    txn3.commit(engine, emptyIterable())
+
+    if (expectedConflict) {
+      // We expect the commit of txn1 to fail because of the conflicting DM actions
+      val ex = intercept[KernelException] {
+        txn1.commit(engine, emptyIterable())
+      }
+      assert(
+        ex.getMessage.contains(
+          "A concurrent writer added a domainMetadata action for the same domain"
+        )
+      )
+    } else {
+      // We expect the commit of txn1 to succeed
+      txn1.commit(engine, emptyIterable())
+      // Verify the final state includes merged domain metadata
+      val expectedMetadata =
+        (winningTxn2DomainMetadatas ++ winningTxn3DomainMetadatas ++ currentTxn1DomainMetadatas)
+          .groupBy(_.getDomain)
+          .mapValues(_.last)
+      assertDomainMetadata(table, engine, expectedMetadata)
+    }
+  }
+
+  test("create table w/o domain metadata") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+
+      // Create an empty table
+      createTxn(engine, tablePath, isNewTable = true, testSchema, Seq.empty)
+        .commit(engine, emptyIterable())
+
+      // Verify that the table doesn't have any domain metadata
+      assertDomainMetadata(table, engine, Map.empty)
+    }
+  }
+
+  test("table w/o domain metadata support fails domain metadata commits") {
```

Comment:
should we also just check that we can read tables with domain metadata in them but with the feature turned off? (slightly orthogonal to changes here but would be good to have a test)

Reply:
I might be missing something here, but I’m not quite sure how this test would work. A table with domain metadata must include “domainMetadata” in its protocol's writer features. When we use the Kernel to read such a table, the feature is automatically turned on (since it’s already implemented). How would we test reading a table with the feature explicitly turned off?

Reply:
ah yea realized it isn't a reader feature so we just always handle the read path - sounds good, disregard!